### DISCUSSION TITLE
How to restrict runner autoscale based of runner labels?

### DISCUSSION THREAD

**Discussion by rahul9170** [actions/actions-runner-controller#1411] [General]
Hello,

Currently the runner autoscales(pull based) based on metric type: **TotalNumberOfQueuedAndInProgressWorkflowRuns** which scales up the runner as soon as it sees a job queued. Can we restrict this so it autoscales based on the runner label mentioned in the job to prevent unnecessary autoscaling?

Thanks,
Rahul Sharma

**Comment by toast-gear** [comment]
Have you tried v0.23.0?

**Reply by rahul9170** [reply]
Yes I am on v0.23.0 only. Is there any parameter I could pass in to runnerdeployment.yaml?

**Reply by toast-gear** [reply]
You shouldn't need to do anything, label support was added in v0.23.0 and having just re-tested it it's working fine in my environment. What does all of your yaml look like? Use code snippets:

**Reply by rahul-FLS** [reply]
Here are the yamls

apiVersion: v1
items:
- apiVersion: actions.summerwind.dev/v1alpha1
  kind: RunnerDeployment
  metadata:
    generation: 3
    labels:
      app: release-27a27040-595d-3961-9348-3e105b625be2-runner
      chart: actions-runner-0.1.0-0.4
      release: release-27a27040-595d-3961-9348-3e105b625be2
    name: actions-runner
    namespace: actions-runner-system
    resourceVersion: "504580"
    uid: 12ae65f3-59d5-490e-a5be-f952e2ab42a5
  spec:
    effectiveTime: null
    replicas: 0
    selector: null
    template:
      metadata: {}
      spec:
        dockerEnabled: false
        dockerdContainerResources: {}
        env:
        - name: RUNNER_FEATURE_FLAG_EPHEMERAL
          value: "true"
        - name: DISABLE_RUNNER_UPDATE
          value: "true"
        ephemeral: true
        group: poc
        image: actionsRunner
        imagePullSecrets:
        - name: secret
        labels:
        - container-runner
        organization: orgName
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: 500m
            memory: 500Mi
        securityContext:
          runAsUser: 1000
        serviceAccountName: actions-runner-sa
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        volumes:
        - emptyDir: {}
          name: tmp


apiVersion: v1
items:
- apiVersion: actions.summerwind.dev/v1alpha1
  kind: HorizontalRunnerAutoscaler
  metadata:
    creationTimestamp: "2022-05-03T11:29:58Z"
    generation: 1
    name: actions-runner-autoscaler
    namespace: actions-runner-system
    resourceVersion: "504533"
    uid: d1c9d7f6-9be2-44c2-83c2-fa2a491f55e3
  spec:
    maxReplicas: 5
    metrics:
    - repositoryNames:
      - repoName
      type: TotalNumberOfQueuedAndInProgressWorkflowRuns
    minReplicas: 0
    scaleTargetRef:
      name: actions-runner

**Reply by toast-gear** [reply]
Format this so it's in a code block, this is incredibly hard to read. What behaviour are you actually seeing? I expected 2 `runner*.yaml` files as I assume you aren't getting only the matching labeled runner scaling?

**Reply by rahul9170** [reply]
Sorry about that. Here are the formatted ones.

**HorizaontalRunnerAutoscaler.yaml**

<img width="620" alt="image" src="https://user-images.githubusercontent.com/61755746/166667558-54d2fe6c-8bac-4ee5-bb4f-923d09c8a0a4.png">


**RunnerDeployment.yaml**

<img width="511" alt="image" src="https://user-images.githubusercontent.com/61755746/166667831-ca39114c-8086-45c4-b374-bf18d0a6c239.png">

I could see runners are getting scaled up if the controller sees any job(irrespective to the job label) queued up in GitHub.

**Reply by toast-gear** [reply]
mmmmm, `container-runner` is unique? Can provide what an offending workflow looks like? What do the controller logs say, I'm espcially interested in the `Suggested desired replicas of $CALCULATED_REPLICA_COUNT by  TotalNumberOfQueuedAndInProgressWorkflowRuns` lines and their surrounding logs. If that label is unique to this bank of runners and you are referencing it in a workflow then I've tested this scenario this morning and it worked OK in my environment so it's slightly perplexing! 🤔

**Comment by mumoshu** [comment]
Could you also share the `kubectl get deployment -o yaml` output for ARC's deployment?

**Comment by rahul9170** [comment]
Hey this seems to be working now not sure what changed. I'll test it more and shout in case of any issue. Thanks again :)

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
